# Volcano Eruptions

Which volcanoes have experienced the longest eruptions?

The dataset `volcanic-eruptions.csv` includes the start and end dates for all volcanic eruptions since 1800, along with each volcano's identification number. For eruptions that are still ongoing as of December 2024, the end date is recorded as December 2024.

This dataset excludes any eruptive pauses shorter than three months. If an eruption resumes after more than three months of inactivity, it is classified as a new eruption.

Note: Volcano Yasur in Vanuatu is not included in this list due to the absence of a clear start date.

In [1]:
# FOR GOOGLE COLAB ONLY.
# Uncomment and run the code below. A dialog will appear to upload files.
# Upload 'volcanic-eruptions.csv' and 'volcano-list.csv'.

# from google.colab import files
# uploaded = files.upload()

In [2]:
import pandas as pd
import matplotlib.pyplot as plt

eruptions = pd.read_csv('volcanic-eruptions.csv')
eruptions.head(3)

,volcano_id,start_date,end_date
0,211020,07-1913,04-1944
1,211020,02-1864,11-1868
2,211020,12-1854,05-1855


### Additional dataset

The dataset `volcano-list.csv` provides detailed information about each volcano, including its name, country, latitude, longitude, and type.

In [3]:
volcanoes = pd.read_csv('volcano-list.csv')
volcanoes.head(3)

,volcano_id,volcano_name,country,volcanic_region_group,volcanic_region,volcano_landform,primary_volcano_type,activity_evidence,last_known_eruption,latitude,longitude,elevation_m,tectonic_setting,dominant_rock_type
0,210010,West Eifel Volcanic Field,Germany,European Volcanic Regions,Central European Volcanic Province,Cluster,Volcanic field,Eruption Dated,8300 BCE,50.170,6.850,600,Rift zone / Continental crust (>25 km),Foidite
1,210020,Chaine des Puys,France,European Volcanic Regions,Western European Volcanic Province,Cluster,Lava dome(s),Eruption Dated,4040 BCE,45.786,2.981,1464,Rift zone / Continental crust (>25 km),Basalt / Picro-Basalt
2,210030,Olot Volcanic Field,Spain,European Volcanic Regions,Western European Volcanic Province,Cluster,Volcanic field,Evidence Credible,Unknown,42.170,2.530,893,Intraplate / Continental crust (>25 km),Trachybasalt / Tephrite Basanite


### Project Ideas:

- Find the volcanoes that were erupting as of Dec 2024.

- Find the volcanoes that have had the longest volcanic eruptions. 

Hints:
- Use `pd.to_datetime`.

- Merge the volcanoes dataframe into eruptions.

- Before the merge, reduce the dataframes to the columns of interest.

- Use `df.sort_values`.

**Task 1: Find the volcanoes that were erupting as of Dec 2024.**

In [4]:
eruptions.info()

<class 'pandas.DataFrame'>
RangeIndex: 3724 entries, 0 to 3723
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   volcano_id  3724 non-null   int64
 1   start_date  3724 non-null   str  
 2   end_date    3724 non-null   str  
dtypes: int64(1), str(2)
memory usage: 87.4 KB


Both the start and end date are in **string format**. Meaning, we have to covert the data to date time.

In [5]:
eruptions['start_date'] = pd.to_datetime(eruptions['start_date'])
eruptions['end_date'] = pd.to_datetime(eruptions['end_date'])
eruptions.info()

<class 'pandas.DataFrame'>
RangeIndex: 3724 entries, 0 to 3723
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   volcano_id  3724 non-null   int64         
 1   start_date  3724 non-null   datetime64[us]
 2   end_date    3724 non-null   datetime64[us]
dtypes: datetime64[us](2), int64(1)
memory usage: 87.4 KB


C:\Users\lenov\AppData\Local\Temp\ipykernel_14648\2555005822.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  eruptions['start_date'] = pd.to_datetime(eruptions['start_date'])
C:\Users\lenov\AppData\Local\Temp\ipykernel_14648\2555005822.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  eruptions['end_date'] = pd.to_datetime(eruptions['end_date'])


In [6]:
#getting the needed data from volcanoes dataset: volcano_id, volcano_name, country
volcano_info = volcanoes[['volcano_id','volcano_name', 'country']]
volcano_info

,volcano_id,volcano_name,country
0,210010,West Eifel Volcanic Field,Germany
1,210020,Chaine des Puys,France
2,210030,Olot Volcanic Field,Spain
3,210040,Calatrava Volcanic Field,Spain
4,211004,Colli Albani,Italy
...,...,...,...
1276,390120,Leskov Island,United Kingdom
1277,390130,Zavodovski,United Kingdom
1278,390140,Protector Seamounts,United Kingdom
1279,390829,Mount Haddington Volcanic Field,Antarctica


In [7]:
#merging the data frames
volcano_analysis = eruptions.merge(volcano_info, on = 'volcano_id', how = 'left')
volcano_analysis

,volcano_id,start_date,end_date,volcano_name,country
0,211020,1913-07-01,1944-04-01,Vesuvius,Italy
1,211020,1864-02-01,1868-11-01,Vesuvius,Italy
2,211020,1854-12-01,1855-05-01,Vesuvius,Italy
3,211020,1855-12-01,1861-12-01,Vesuvius,Italy
4,211020,1824-07-01,1834-09-01,Vesuvius,Italy
...,...,...,...,...,...
3719,390090,1999-01-01,1999-01-01,Saunders,United Kingdom
3720,390090,2014-11-01,2024-10-01,Saunders,United Kingdom
3721,390090,1995-04-01,1998-02-01,Saunders,United Kingdom
3722,390130,2016-03-01,2016-06-01,Zavodovski,United Kingdom


In [8]:
# Finally, we can look for the volcanoes that erupted in 2024. 
volcano_2024 = volcano_analysis.query('end_date >= "2024-12-01"')
volcano_2024

,volcano_id,start_date,end_date,volcano_name,country
12,211040,1934-02-01,2024-12-01,Stromboli,Italy
25,211060,2022-11-01,2024-12-01,Etna,Italy
97,221080,1967-07-01,2024-12-01,Erta Ale,Ethiopia
111,222120,2017-04-01,2024-12-01,"Lengai, Ol Doinyo",Tanzania
134,223020,2018-04-01,2024-12-01,Nyamulagira,DR Congo
322,234010,2012-09-01,2024-12-01,Heard,Australia
339,241040,2024-05-01,2024-12-01,Whakaari/White Island,New Zealand
488,243060,2015-10-01,2024-12-01,Tofua,Tonga
499,243080,2024-12-01,2024-12-01,Home Reef,Tonga
550,251020,2018-05-01,2024-12-01,Manam,Papua New Guinea


**Task 2: Find the volcanoes that have had the longest volcanic eruptions.**

In [ ]:
#We can compute for the number of days each volcano erupted by subtracting the start date from the end date.
volcano_analysis['days'] = ((volcano_analysis['end_date'] - volcano_analysis['start_date']).dt.days)
volcano_analysis

,volcano_id,start_date,end_date,volcano_name,country,days
0,211020,1913-07-01,1944-04-01,Vesuvius,Italy,11232
1,211020,1864-02-01,1868-11-01,Vesuvius,Italy,1735
2,211020,1854-12-01,1855-05-01,Vesuvius,Italy,151
3,211020,1855-12-01,1861-12-01,Vesuvius,Italy,2192
4,211020,1824-07-01,1834-09-01,Vesuvius,Italy,3714
...,...,...,...,...,...,...
3719,390090,1999-01-01,1999-01-01,Saunders,United Kingdom,0
3720,390090,2014-11-01,2024-10-01,Saunders,United Kingdom,3622
3721,390090,1995-04-01,1998-02-01,Saunders,United Kingdom,1037
3722,390130,2016-03-01,2016-06-01,Zavodovski,United Kingdom,92


In [10]:
#We can sort the data by "days" and get the top 10 volcanoes that erupted for the longest time.
top_10_longest_eruptions = volcano_analysis.sort_values(by = 'days', ascending = False).head(10)
top_10_longest_eruptions

,volcano_id,start_date,end_date,volcano_name,country,days
2932,342030,1922-06-01,2024-12-01,Santa Maria,Guatemala,37439
1575,268010,1933-08-01,2024-12-01,Dukono,Indonesia,33360
12,211040,1934-02-01,2024-12-01,Stromboli,Italy,33176
3362,352090,1934-08-01,2011-03-01,Sangay,Ecuador,27971
2782,332010,1823-02-01,1894-12-01,Kilauea,United States,26236
1813,282080,1955-10-01,2016-08-01,Aira,Japan,22220
97,221080,1967-07-01,2024-12-01,Erta Ale,Ethiopia,20973
3708,390020,1972-12-01,2024-12-01,Erebus,Antarctica,18993
164,223030,1927-03-01,1977-01-01,Nyiragongo,DR Congo,18204
655,253030,1890-01-01,1935-01-01,Victory,Papua New Guinea,16435


**ANSWER:** Based to the results, the Santa Maria Volcano in Guatemala erupted for the longest time, which lasted for 37439 days. 